cmd/snap-bootstrap/cmd_initramfs_mounts.go

```diff
@@ -0,0 +1,201 @@
+)
+
+// generateMountsMode* is called multiple times from initramfs until it
+// no longer generates more mount points and just returns an empty output.
+func generateMountsModeInstall() error {
+	// 1. always ensure seed partition is mounted
+	isMounted, err := osutilIsMounted(filepath.Join(runMnt, "ubuntu-seed"))
+	if err != nil {
+		return err
+	}
+	if !isMounted {
+		fmt.Fprintf(stdout, "/dev/disk/by-label/ubuntu-seed %s\n", filepath.Join(runMnt, "ubuntu-seed"))
+		return nil
+	}
+	// XXX: how do we select a different recover system from the cmdline?
+
+	// 2. (auto) select recovery system for now
+	isMounted, err = osutilIsMounted(filepath.Join(runMnt, "base"))
+	if err != nil {
+		return err
+	}
+	if !isMounted {
+		l, err := filepath.Glob(filepath.Join(runMnt, "/ubuntu-seed/systems/*"))
+		if err != nil {
+			return err
+		}
+		if len(l) == 0 {
+			return fmt.Errorf("cannot find a recovery system")
+		}
+		if len(l) > 1 {
+			return fmt.Errorf("cannot use multiple recovery systems yet")
+		}
+		// load the seed and generate mounts for kernel/base
+		label := filepath.Base(l[0])
+		seedDir := filepath.Dir(filepath.Dir(l[0]))
+		deviceSeed, err := seed.Open(seedDir, label)
+		if err != nil {
+			return err
+		}
+		// load assertions into a temporary database
+		if err := deviceSeed.LoadAssertions(nil, nil); err != nil {
+			return err
+		}
+		perf := timings.New(nil)
+		// XXX: LoadMeta will verify all the snaps in the
+		// seed, that is probably too much. We can expose more
+		// dedicated helpers for this later.
+		if err := deviceSeed.LoadMeta(perf); err != nil {
+			return err
+		}
+		// XXX: do we need more cross checks here?
+		for _, essentialSnap := range deviceSeed.EssentialSnaps() {
+			snapf, err := snap.Open(essentialSnap.Path)
+			if err != nil {
+				return err
+			}
+			info, err := snap.ReadInfoFromSnapFile(snapf, essentialSnap.SideInfo)
+			if err != nil {
+				return err
+			}
+			switch info.GetType() {
+			case snap.TypeBase:
+				fmt.Fprintf(stdout, "%s %s\n", essentialSnap.Path, filepath.Join(runMnt, "base"))
+			case snap.TypeKernel:
+				fmt.Fprintf(stdout, "%s %s\n", essentialSnap.Path, filepath.Join(runMnt, "kernel"))
+			}
+		}
+
+		return nil
+	}
+
+	// 3. mount "ubuntu-data" on a tmpfs
+	isMounted, err = osutilIsMounted(filepath.Join(runMnt, "ubuntu-data"))
+	if err != nil {
+		return err
+	}
+	if !isMounted {
+		// XXX: is there a better way?
+		fmt.Fprintf(stdout, "--type=tmpfs tmpfs /run/mnt/ubuntu-data\n")
+		return nil
+	}
+
+	// 4. done, no output, no error indicates to initramfs we are done
+	//    with mounting stuff
+	return nil
+}
+
+func generateMountsModeRecover() error {
+	return fmt.Errorf("recover mode mount generation not implemented yet")
+}
+
+func generateMountsModeRun() error {
+	return fmt.Errorf("run mode mount generation not implemented yet")
+}
+
+func isInstallMode(content []byte) bool {
+	// XXX: deal with whitespace
+	if bytes.Contains(content, []byte("snapd_recovery_mode=install")) {
```

Comment:
this will also work with "install-foo", do we want that?

in general I would slightly prefer a single  whichMode(content) string function ? instead of all the helpers